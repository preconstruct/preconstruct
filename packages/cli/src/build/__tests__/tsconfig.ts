import path from "path";
import build from "..";
import {
  snapshotDirectory,
  testdir,
  ts,
  repoNodeModules,
} from "../../../test-utils";

jest.setTimeout(6000000);

jest.mock("../../prompt");

test("should build with custom tsconfig filename in standalone package", async () => {
  let tmpPath = await testdir({
    "package.json": JSON.stringify({
      name: "typescript-custom-config",
      main: "dist/typescript-custom-config.cjs.js",
      module: "dist/typescript-custom-config.esm.js",
      dependencies: {
        "@babel/runtime": "^7.8.7",
      },
      devDependencies: {
        typescript: "^3.8.3",
      },
      preconstruct: {
        tsconfig: "tsconfig.build.json",
      },
    }),
    ".babelrc": JSON.stringify({
      presets: [require.resolve("@babel/preset-typescript")],
    }),
    node_modules: {
      kind: "symlink",
      path: repoNodeModules,
    },
    "tsconfig.build.json": JSON.stringify({
      compilerOptions: {
        module: "commonjs",
        moduleResolution: "node",
        jsx: "react",
        declaration: true,
        isolatedModules: true,
        strict: true,
        esModuleInterop: true,
        noEmit: true,
      },
    }),
    "src/index.ts": ts`
      export const thing = "custom-config" as const;
    `,
  });
  await build(tmpPath);
  await snapshotDirectory(path.join(tmpPath, "dist"), { files: "all" });
});

test("should fail to build with invalid custom tsconfig filename in standalone package", async () => {
  let tmpPath = await testdir({
    "package.json": JSON.stringify({
      name: "typescript-invalid-custom-config",
      main: "dist/typescript-invalid-custom-config.cjs.js",
      module: "dist/typescript-invalid-custom-config.esm.js",
      dependencies: {
        "@babel/runtime": "^7.8.7",
      },
      devDependencies: {
        typescript: "^3.8.3",
      },
      preconstruct: {
        tsconfig: "tsconfig.invalid.json",
      },
    }),
    ".babelrc": JSON.stringify({
      presets: [require.resolve("@babel/preset-typescript")],
    }),
    node_modules: {
      kind: "symlink",
      path: repoNodeModules,
    },
    "tsconfig.build.json": JSON.stringify({
      compilerOptions: {
        target: "esnext",
        module: "esnext",
        jsx: "react",
        isolatedModules: true,
        strict: true,
        moduleResolution: "node",
        esModuleInterop: true,
        noEmit: true,
      },
      "tsconfig.json": JSON.stringify({
        compilerOptions: {
          module: "nodenext",
          moduleResolution: "nodenext",
          jsx: "react",
          declaration: true,
          isolatedModules: true,
          strict: true,
          esModuleInterop: true,
          noEmit: true,
        },
      }),
    }),
    "src/index.ts": ts`
      export const thing = "custom-invalid-config" as const;
    `,
  });
  await expect(build(tmpPath)).rejects.toMatchInlineSnapshot(
    `[Error: üéÅ typescript-invalid-custom-config the custom TypeScript config file tsconfig.invalid.json does not exist.]`
  );
});

test("should build with custom tsconfig filename in monorepo package", async () => {
  let tmpPath = await testdir({
    "package.json": JSON.stringify({
      name: "monorepo-root",
      private: true,
      workspaces: ["packages/*"],
      devDependencies: {
        typescript: "^3.8.3",
      },
      preconstruct: {
        packages: ["packages/*"],
      },
    }),
    ".babelrc": JSON.stringify({
      presets: [require.resolve("@babel/preset-typescript")],
    }),
    node_modules: {
      kind: "symlink",
      path: repoNodeModules,
    },
    "packages/pkg-a/package.json": JSON.stringify({
      name: "pkg-a",
      main: "dist/pkg-a.cjs.js",
      module: "dist/pkg-a.esm.js",
      preconstruct: {
        tsconfig: "tsconfig.build.json",
      },
    }),
    "packages/pkg-a/tsconfig.build.json": JSON.stringify({
      compilerOptions: {
        module: "commonjs",
        jsx: "react",
        declaration: true,
        isolatedModules: true,
        strict: true,
        moduleResolution: "node",
        esModuleInterop: true,
        noEmit: true,
      },
    }),
    "packages/pkg-a/src/index.ts": ts`
      export const thing = "monorepo-custom-config";
    `,
  });
  await build(tmpPath);
  await snapshotDirectory(path.join(tmpPath, "packages/pkg-a/dist"), {
    files: "all",
  });
});

test("should build a monorepo package with custom tsconfig located in a relative folder", async () => {
  let tmpPath = await testdir({
    "package.json": JSON.stringify({
      name: "monorepo-root",
      private: true,
      workspaces: ["packages/*"],
      devDependencies: {
        typescript: "^3.8.3",
      },
      preconstruct: {
        packages: ["packages/*"],
      },
    }),
    ".babelrc": JSON.stringify({
      presets: [require.resolve("@babel/preset-typescript")],
    }),
    node_modules: {
      kind: "symlink",
      path: repoNodeModules,
    },
    "packages/pkg-a/package.json": JSON.stringify({
      name: "pkg-a",
      main: "dist/pkg-a.cjs.js",
      module: "dist/pkg-a.esm.js",
      preconstruct: {
        tsconfig: "./folder/tsconfig.json",
      },
    }),
    "packages/pkg-a/folder/tsconfig.json": JSON.stringify({
      compilerOptions: {
        module: "commonjs",
        jsx: "react",
        declaration: true,
        isolatedModules: true,
        strict: true,
        moduleResolution: "node",
        esModuleInterop: true,
        noEmit: true,
      },
      include: ["../src/**/*"],
    }),
    "packages/pkg-a/src/index.ts": ts`
      export const thing = "monorepo-custom-config";
    `,
  });
  await build(tmpPath);
  await snapshotDirectory(path.join(tmpPath, "packages/pkg-a/dist"), {
    files: "all",
  });
});

test("should fail to build a monorepo package with relative folder and invalid typescript configuration inclusions rules", async () => {
  let tmpPath = await testdir({
    "package.json": JSON.stringify({
      name: "monorepo-root",
      private: true,
      workspaces: ["packages/*"],
      devDependencies: {
        typescript: "^3.8.3",
      },
      preconstruct: {
        packages: ["packages/*"],
      },
    }),
    ".babelrc": JSON.stringify({
      presets: [require.resolve("@babel/preset-typescript")],
    }),
    node_modules: {
      kind: "symlink",
      path: repoNodeModules,
    },
    "packages/pkg-a/package.json": JSON.stringify({
      name: "pkg-a",
      main: "dist/pkg-a.cjs.js",
      module: "dist/pkg-a.esm.js",
      preconstruct: {
        tsconfig: "./folder/tsconfig.json",
      },
    }),
    "packages/pkg-a/folder/tsconfig.json": JSON.stringify({
      compilerOptions: {
        module: "commonjs",
        jsx: "react",
        declaration: true,
        isolatedModules: true,
        strict: true,
        moduleResolution: "node",
        esModuleInterop: true,
        noEmit: true,
      },
    }),
    "packages/pkg-a/src/index.ts": ts`
      export const thing = "monorepo-custom-config";
    `,
  });
  await expect(build(tmpPath)).rejects.toThrowError();
});

test("should fail to build with invalid custom tsconfig filename in monorepo package", async () => {
  let tmpPath = await testdir({
    "package.json": JSON.stringify({
      name: "monorepo-root",
      private: true,
      workspaces: ["packages/*"],
      devDependencies: {
        typescript: "^3.8.3",
      },
      preconstruct: {
        packages: ["packages/*"],
      },
    }),
    ".babelrc": JSON.stringify({
      presets: [require.resolve("@babel/preset-typescript")],
    }),
    node_modules: {
      kind: "symlink",
      path: repoNodeModules,
    },
    "packages/pkg-a/package.json": JSON.stringify({
      name: "pkg-a",
      main: "dist/pkg-a.cjs.js",
      module: "dist/pkg-a.esm.js",
      preconstruct: {
        tsconfig: "tsconfig.invalid.json",
      },
    }),
    "packages/pkg-a/tsconfig.json": JSON.stringify({
      compilerOptions: {
        target: "esnext",
        module: "esnext",
        jsx: "react",
        isolatedModules: true,
        strict: true,
        moduleResolution: "node",
        esModuleInterop: true,
        noEmit: true,
      },
    }),
    "packages/pkg-a/tsconfig.build.json": JSON.stringify({
      compilerOptions: {
        module: "commonjs",
        moduleResolution: "node",
        jsx: "react",
        declaration: true,
        isolatedModules: true,
        strict: true,
        esModuleInterop: true,
        noEmit: true,
      },
    }),
    "packages/pkg-a/src/index.ts": ts`
      export const thing = "custom-invalid-config";
    `,
  });
  await expect(build(tmpPath)).rejects.toMatchInlineSnapshot(
    `[Error: üéÅ pkg-a the custom TypeScript config file tsconfig.invalid.json does not exist.]`
  );
});
