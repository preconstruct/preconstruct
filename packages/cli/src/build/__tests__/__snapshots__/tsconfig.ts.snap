// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should build a monorepo package with custom tsconfig located in a relative folder: declarations/src/index.d.ts 1`] = `
"export declare const thing = "monorepo-custom-config";
"
`;

exports[`should build a monorepo package with custom tsconfig located in a relative folder: pkg-a.cjs.d.ts 1`] = `
"export * from "./declarations/src/index.js";
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoicGtnLWEuY2pzLmQudHMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyIuL2RlY2xhcmF0aW9ucy9zcmMvaW5kZXguZC50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQSJ9
"
`;

exports[`should build a monorepo package with custom tsconfig located in a relative folder: pkg-a.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const thing = "monorepo-custom-config";

exports.thing = thing;
"
`;

exports[`should build a monorepo package with custom tsconfig located in a relative folder: pkg-a.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === "production") {
  module.exports = require("./pkg-a.cjs.prod.js");
} else {
  module.exports = require("./pkg-a.cjs.dev.js");
}
"
`;

exports[`should build a monorepo package with custom tsconfig located in a relative folder: pkg-a.cjs.prod.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const thing = "monorepo-custom-config";

exports.thing = thing;
"
`;

exports[`should build a monorepo package with custom tsconfig located in a relative folder: pkg-a.esm.js 1`] = `
"const thing = "monorepo-custom-config";

export { thing };
"
`;

exports[`should build with custom tsconfig filename in monorepo package: declarations/src/index.d.ts 1`] = `
"export declare const thing = "monorepo-custom-config";
"
`;

exports[`should build with custom tsconfig filename in monorepo package: pkg-a.cjs.d.ts 1`] = `
"export * from "./declarations/src/index.js";
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoicGtnLWEuY2pzLmQudHMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyIuL2RlY2xhcmF0aW9ucy9zcmMvaW5kZXguZC50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQSJ9
"
`;

exports[`should build with custom tsconfig filename in monorepo package: pkg-a.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const thing = "monorepo-custom-config";

exports.thing = thing;
"
`;

exports[`should build with custom tsconfig filename in monorepo package: pkg-a.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === "production") {
  module.exports = require("./pkg-a.cjs.prod.js");
} else {
  module.exports = require("./pkg-a.cjs.dev.js");
}
"
`;

exports[`should build with custom tsconfig filename in monorepo package: pkg-a.cjs.prod.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const thing = "monorepo-custom-config";

exports.thing = thing;
"
`;

exports[`should build with custom tsconfig filename in monorepo package: pkg-a.esm.js 1`] = `
"const thing = "monorepo-custom-config";

export { thing };
"
`;

exports[`should build with custom tsconfig filename in standalone package: declarations/src/index.d.ts 1`] = `
"export declare const thing: "custom-config";
"
`;

exports[`should build with custom tsconfig filename in standalone package: typescript-custom-config.cjs.d.ts 1`] = `
"export * from "./declarations/src/index.js";
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoidHlwZXNjcmlwdC1jdXN0b20tY29uZmlnLmNqcy5kLnRzIiwic291cmNlUm9vdCI6IiIsInNvdXJjZXMiOlsiLi9kZWNsYXJhdGlvbnMvc3JjL2luZGV4LmQudHMiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUEifQ==
"
`;

exports[`should build with custom tsconfig filename in standalone package: typescript-custom-config.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const thing = "custom-config";

exports.thing = thing;
"
`;

exports[`should build with custom tsconfig filename in standalone package: typescript-custom-config.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === "production") {
  module.exports = require("./typescript-custom-config.cjs.prod.js");
} else {
  module.exports = require("./typescript-custom-config.cjs.dev.js");
}
"
`;

exports[`should build with custom tsconfig filename in standalone package: typescript-custom-config.cjs.prod.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const thing = "custom-config";

exports.thing = thing;
"
`;

exports[`should build with custom tsconfig filename in standalone package: typescript-custom-config.esm.js 1`] = `
"const thing = "custom-config";

export { thing };
"
`;
